src/test/object-util_test.ts

```diff
@@ -1,137 +1,46 @@
-    await t.test(`cannot set ${key}`, () => {
-      const obj = {};
-      setDeepValue(obj, [key], 123);
-      assert.deepEqual(obj, {});
+  await t.test('existing object if value last key already exists', () => {
+    assert.deepEqual(getDeepObject({foo: {bar: true}}, 'foo', 'baz'), {
+      bar: true
     });
-
-    await t.test('cannot set proto deeply', () => {
-      const obj = {foo: {}};
-      setDeepValue(obj, ['foo', key], 123);
-      assert.deepEqual(obj, {foo: {}});
-    });
-  }
-
-  await t.test('sets top level key', () => {
-    const obj: Record<PropertyKey, unknown> = {};
-    setDeepValue(obj, ['foo'], 303);
-    assert.deepEqual(obj, {foo: 303});
   });
 
-  await t.test('sets deep key', () => {
-    const obj: Record<PropertyKey, unknown> = {foo: {}};
-    setDeepValue(obj, ['foo', 'bar'], 303);
-    assert.deepEqual(obj, {foo: {bar: 303}});
-  });
+  const disallowedNames = ['__proto__', 'constructor', 'prototype'];
 
-  await t.test('replaces null object value with object', () => {
-    const obj: Record<PropertyKey, unknown> = {foo: null};
-    setDeepValue(obj, ['foo', 'bar'], 303);
-    assert.deepEqual(obj, {
-      foo: {
-        bar: 303
-      }
+  for (const name of disallowedNames) {
+    await t.test(`setting disallowed ${name} returns object as-is`, () => {
+      assert.deepEqual(getDeepObject({foo: 'bar'}, name, 'foo'), {foo: 'bar'});
     });
-  });
-
-  await t.test('replaces null array value with array', () => {
-    const obj: Record<PropertyKey, unknown> = {foo: null};
-    setDeepValue(obj, ['foo', 0], 303);
-    assert.deepEqual(obj, {
-      foo: [303]
-    });
-  });
-
-  await t.test('creates new objects for object values', () => {
-    const obj: Record<PropertyKey, unknown> = {};
-    setDeepValue(obj, ['foo', 'bar'], 303);
-    assert.deepEqual(obj, {
-      foo: {
-        bar: 303
-      }
-    });
-  });
+  }
 
-  await t.test('creates new arrays for array values', () => {
-    const obj: Record<PropertyKey, unknown> = {};
-    setDeepValue(obj, ['foo', 0], 303);
-    assert.deepEqual(obj, {
-      foo: [303]
+  await t.test('can key into existing sub-object', () => {
+    assert.deepEqual(getDeepObject({foo: {bar: 'baz'}}, 'foo', ''), {
+      bar: 'baz'
     });
   });
 
-  await t.test('creates new arrays with string indices', () => {
-    const obj: Record<PropertyKey, unknown> = {};
-    setDeepValue(obj, ['foo', '0'], 303);
-    assert.deepEqual(obj, {
-      foo: [303]
-    });
+  await t.test('replaces null with new object', () => {
+    assert.deepEqual(getDeepObject({foo: null}, 'foo', ''), {});
   });
 
-  await t.test('treats decimal strings as regular keys', () => {
-    const obj: Record<PropertyKey, unknown> = {};
-    setDeepValue(obj, ['foo', '10.0'], 303);
-    assert.deepEqual(obj, {
-      foo: {
-        '10.0': 303
-      }
-    });
+  await t.test('treats decimals as object keys', () => {
+    assert.deepEqual(getDeepObject({}, 'foo', '1.5'), {'1.5': {}});
```

Comment:
the ci failure is probably this one since it'll deep equal `{}` (i.e. `obj['foo']` will be set to a new object)

Reply:
Yeah, I realized